### PR TITLE
Make segment column resizeable

### LINKED ISSUE #183: Resizable columns for table view
Proposals:
- ability to resize left part of the table view
- strong visual separation between dimensions and measures cells

### PR DESCRIPTION
**Pull request by cedrics**
Fixes #183 

That was pretty straight forward! Maybe too easy, pretty sure I missed something 😛 

The original ticket wanted a "strong visual separation between dimensions and measures cells" probably makes more sense than the weird grey bar I added 😸 but I haven't done it yet. This probably just means a bar dividing the segment column and the others?

I like to change the `ResizeHandle` and make the `value` a controlled property instead. The state is currently kept in the component rendering the `ResizeHandle` as well as the `ResizeHandle` itself, which can easily lead to bugs like the one I describe in ef05a96 (in this case due to handle being rerendered but the component rendering it is not). WDYT?

resize-column

### CODE REVIEW

File: src/client/components/resize-handle/resize-handle.tsx [hunk 1]
```diff
@@ -118,7 +120,7 @@ export class ResizeHandle extends React.Component<ResizeHandleProps, ResizeHandl
       style={style}
       onMouseDown={this.onMouseDown}
     >
-      <SvgIcon svg={require("../../icons/drag-handle.svg")} />
+      {children === undefined ? <SvgIcon svg={require("../../icons/drag-handle.svg")} /> : children}
```
Comment (adrianmroz) [inline src/client/components/resize-handle/resize-handle.tsx]: `cchildren || <SvgIcon>`?

File: src/client/visualizations/table/table.scss [hunk 2]
```diff
@@ -91,7 +90,7 @@ $header-padding-top: 12px;
       background-color: $white;
       left: 0;
       top: 0;
-      width: $segment-width - $space-left;
+      width: calc(100% - #{$space-left});
```
Comment (adrianmroz) [inline src/client/visualizations/table/table.scss]: How about browser support?

File: src/client/visualizations/table/table.tsx [hunk 3]
```diff
@@ -452,8 +464,17 @@ export class Table extends BaseVisualization<TableState> {
...
+    const defaultSegmentWidth = this.defaultSegmentWidth();
     const corner = <div className="corner">
       <div className="corner-wrap">{segmentTitle}</div>
+      <ResizeHandle
+        direction={Direction.LEFT}
+        onResize={this.setSegmentWidth}
+        min={defaultSegmentWidth}
+        max={MAX_SEGMENT_WIDTH}
+        initialValue={segmentWidth || defaultSegmentWidth}
+        children={null}
```
Comment (adrianmroz) [inline src/client/visualizations/table/table.tsx]: You pass null and check for undefined. Scary ;)

File: src/client/visualizations/table/table.tsx [hunk 4]
```diff
@@ -50,6 +51,7 @@ const ROW_HEIGHT = 30;
 const SPACE_LEFT = 10;
 const SPACE_RIGHT = 10;
 const HIGHLIGHT_BUBBLE_V_OFFSET = -4;
+const MAX_SEGMENT_WIDTH = Number.MAX_SAFE_INTEGER;
```
Comment (adrianmroz) [inline src/client/visualizations/table/table.tsx]: Probably we should set there something like table width.

**Comment by adrianmroz** [comment]
Not sure if I follow, but you want to get rid of state in ResizeHandle and give it value and onVaalueChage props? I think it is a way to go if possible.

File: src/client/visualizations/table/table.scss (hunk 2 above)
Comment (cedrics) [reply src/client/visualizations/table/table.scss]: Calc is available in all modern browsers (See [mdn](https://developer.mozilla.org/de/docs/Web/CSS/calc#Browser_Kompatibilit%C3%A4t)). If you prefer I can set the width in JS, doesn't really matter to me 😸

File: src/client/visualizations/table/table.tsx (hunk 3 above)
Comment (cedrics) [reply src/client/visualizations/table/table.tsx]: Yeah 😛 As the only use case is to not render the icon, I change it to a flag to skip it.

**Comment by cedrics** [comment]
> Not sure if I follow, but you want to get rid of state in ResizeHandle and give it value and onVaalueChage props? I think it is a way to go if possible.

Yes, that's what I meant 👍

File: src/client/components/dimension-measure-panel/dimension-measure-panel.tsx [hunk 5]
```diff
@@ -117,11 +117,12 @@ export class DimensionMeasurePanel extends React.Component<DimensionMeasurePanel
           style={dimensionListStyle}
         />
         {showResizeHandle && <ResizeHandle
+          renderIcon={true}
```
Comment (adrianmroz) [inline src/client/components/dimension-measure-panel/dimension-measure-panel.tsx]: Idea for consideration: maybe add prop `as` and in this case pass `<SvgIcon>`?

Comment (cedrics) [reply src/client/components/dimension-measure-panel/dimension-measure-panel.tsx]: Oh, yeah you are right. Makes much more sense just to render the icon as a child 😸

File: src/client/visualizations/table/table.tsx (hunk 4 above)
Comment (cedrics) [reply src/client/visualizations/table/table.tsx]: I will change this. You can actually move the column of the screen and then can't grab it again. Great catch!

File: src/client/components/dimension-measure-panel/dimension-measure-panel.tsx (hunk 5 above)
Comment (adrianmroz) [reply src/client/components/dimension-measure-panel/dimension-measure-panel.tsx]: `children` is even better!

File: src/client/visualizations/table/table.scss (hunk 2 above)
Comment (cedrics) [reply src/client/visualizations/table/table.scss]: Didn't see that before, but it is already used in the code base. See https://github.com/allegro/turnilo/search?q=calc

**Comment by cedrics** [comment]
@adrianmroz could you take another look? I updated the PR. Thank you!

**Comment by cedrics** [comment]
Ok, thanks for the review! Let me know when you are ready to merge it or just hit the button yourself 😄

**Comment by grzaks** [comment]
Looks very useful. When is this planned to be merged with master branch?

**Pull request merged**